From 420957e4c56f65703c6f2f24da0ea35c6b7bbcda Mon Sep 17 00:00:00 2001
From: Stefano Brivio <sbrivio@redhat.com>
Date: Thu, 27 Jul 2017 21:52:30 +0200
Subject: [PATCH] utils: return default family when rtm_family is not
 RTNL_FAMILY_IPMR/IP6MR

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1475762
Upstream Status: iproute2.git commit 5ce897a03bfd

commit 5ce897a03bfda76dc66dc1acfa014fc0e3d3022a
Author: Hangbin Liu <liuhangbin@gmail.com>
Date:   Thu Jul 27 17:44:15 2017 +0800

    utils: return default family when rtm_family is not RTNL_FAMILY_IPMR/IP6MR

    When we get a multicast route, the rtm_type is RTN_MULTICAST, but the
    rtm_family may be AF_INET. If we only check the type with RTNL_FAMILY_IPMR,
    we will get malformed address. e.g.

    + ip -4 route add multicast 172.111.1.1 dev em1 table main

    Before fix:
    + ip route list type multicast table main
    multicast ac6f:101:800:400:400:0:3c00:0 dev em1 scope link

    After fix:
    + ip route list type multicast table main
    multicast 172.111.1.1 dev em1 scope link

    Fixes: 56e3eb4c3400 ("ip: route: fix multicast route dumps")
    Signed-off-by: Hangbin Liu <liuhangbin@gmail.com>
    Acked-by: Phil Sutter <phil@nwl.cc>

Signed-off-by: Stefano Brivio <sbrivio@redhat.com>
---
 lib/utils.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/lib/utils.c b/lib/utils.c
index 7d6ee53..9f55391 100644
--- a/lib/utils.c
+++ b/lib/utils.c
@@ -1219,5 +1219,11 @@ int get_real_family(int rtm_type, int rtm_family)
 	if (rtm_type != RTN_MULTICAST)
 		return rtm_family;
 
-	return rtm_family == RTNL_FAMILY_IPMR ? AF_INET : AF_INET6;
+	if (rtm_family == RTNL_FAMILY_IPMR)
+		return AF_INET;
+
+	if (rtm_family == RTNL_FAMILY_IP6MR)
+		return AF_INET6;
+
+	return rtm_family;
 }
-- 
1.8.3.1

